File: Data Structures & Algorithms/two-integer-sum/submission-4.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        excess = {}
        for i in range(len(nums)):
            diff = target - nums[i]
            if diff in excess:
                return [excess.get(diff, 0), i]
            excess[nums[i]] = i

File: Data Structures & Algorithms/three-integer-sum/submission-4.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // nested twosum
        sort(nums.begin(), nums.end());
        set<vector<int>> numset;

        for (int i = 0; i < nums.size(); i++) {
            int start = i + 1;
            int end = nums.size() - 1;
            while (start < end) {
                int sum = nums[start] + nums[end] + nums[i];
                if (sum == 0) {
                    numset.insert({nums[start], nums[end], nums[i]});
                    start++;
                }
                else if (sum < 0) start++;
                else end--;
            }
        }
        return {numset.begin(), numset.end()};
    }
};
